File: core/src/test/kotlin/com/pokerface/core/EquityTest.kt
package com.pokerface.core

import kotlin.random.Random
import kotlin.test.Test
import kotlin.test.assertTrue

class EquityTest {

    private fun cards(vararg codes: String) = codes.map { Card.fromCode(it) }

    @Test
    fun `pocket aces preflop is a big favorite`() {
        val result = EquityCalculator.estimate(
            heroHole = cards("As", "Ah"),
            board = emptyList(),
            trials = 8000,
            random = Random(42),
        )
        // AA vs random hand heads-up is ~85%.
        assertTrue(result.equity in 0.78..0.92, "unexpected equity: ${result.equity}")
    }

    @Test
    fun `nut flush on river is near certain`() {
        val result = EquityCalculator.estimate(
            heroHole = cards("As", "Ks"),
            board = cards("2s", "7s", "9s", "3d", "4h"),
        )
        assertTrue(result.equity > 0.90, "unexpected equity: ${result.equity}")
    }

    @Test
    fun `worst hand on river with no outs is near zero`() {
        // Board makes a straight on the board itself is tricky to guarantee "worst",
        // instead check a clearly weak two-seven offsuit style low card vs a strong board scenario indirectly
        // by verifying probabilities sum to 1.
        val result = EquityCalculator.estimate(
            heroHole = cards("2c", "7d"),
            board = cards("As", "Ks", "Qs", "Js", "10h"),
        )
        val total = result.winProbability + result.tieProbability + result.loseProbability
        assertTrue(total in 0.999..1.001)
    }

    @Test
    fun `probabilities always sum to one preflop`() {
        val result = EquityCalculator.estimate(
            heroHole = cards("7h", "2c"),
            board = emptyList(),
            trials = 4000,
            random = Random(7),
        )
        val total = result.winProbability + result.tieProbability + result.loseProbability
        assertTrue(total in 0.999..1.001)
    }
}

File: web/src/poker/__tests__/equity.test.ts
import { describe, expect, it } from "vitest";
import { cardFromCode } from "../card";
import { equityOf, estimateEquity } from "../equity";

const cards = (...codes: string[]) => codes.map(cardFromCode);

function seededRng(seed: number): () => number {
  let s = seed;
  return () => {
    s = (s * 1103515245 + 12345) & 0x7fffffff;
    return s / 0x80000000;
  };
}

describe("EquityCalculator", () => {
  it("pocket aces preflop is a big favorite", () => {
    const result = estimateEquity(cards("As", "Ah"), [], 6000, seededRng(42));
    const equity = equityOf(result);
    expect(equity).toBeGreaterThan(0.75);
    expect(equity).toBeLessThan(0.94);
  });

  it("nut flush on river is near certain", () => {
    const result = estimateEquity(cards("As", "Ks"), cards("2s", "7s", "9s", "3d", "4h"));
    expect(equityOf(result)).toBeGreaterThan(0.9);
  });

  it("probabilities always sum to one on the river", () => {
    const result = estimateEquity(cards("2c", "7d"), cards("As", "Ks", "Qs", "Js", "10h"));
    const total = result.winProbability + result.tieProbability + result.loseProbability;
    expect(total).toBeCloseTo(1, 3);
  });

  it("probabilities always sum to one preflop", () => {
    const result = estimateEquity(cards("7h", "2c"), [], 3000, seededRng(7));
    const total = result.winProbability + result.tieProbability + result.loseProbability;
    expect(total).toBeCloseTo(1, 3);
  });
});
